modules/twitter/src/main/java/org/apache/ignite/stream/twitter/TwitterStreamer.java

```diff
@@ -0,0 +1,280 @@
+import org.apache.ignite.internal.util.typedef.internal.A;
+import org.apache.ignite.stream.StreamAdapter;
+import twitter4j.Status;
+import twitter4j.TwitterObjectFactory;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Streamer that consumes from a Twitter Streaming API and feeds transformed key-value pairs, by default < tweetId, text>,
+ * into an {@link IgniteDataStreamer} instance.
+ * <p>
+ * This streamer uses https://dev.twitter.com/streaming API and supports Public API, User Streams, Site Streams and Firehose.
+ * <p>
+ * You can also provide a {@link TweetTransformer} to convert the incoming message into cache entries to override
+ * default transformer.
+ * <p>
+ * This Streamer features:
+ *
+ * <ul>
+ *     <li>Supports OAuth1 authentication scheme. <br/> BasicAuth not supported by Streaming API https://dev.twitter.com/streaming/overview/connecting</li>
+ *     <li>Provide all params in apiParams map. https://dev.twitter.com/streaming/overview/request-parameters</li>
+ * </ul>
+ *
+ */
+public class TwitterStreamer<K, V> extends StreamAdapter<String, K, V> {
+
+    /** Logger. */
+    private IgniteLogger log;
+
+    /**
+     * The message transformer that converts an incoming Tweet into cache entries.
+     * If not provided default transformer will be used.
+     */
+    private TweetTransformer<K, V> transformer;
+
+    /** Twitter Streaming API params. See https://dev.twitter.com/streaming/overview/request-parameters */
+    private Map<String, String> apiParams = new HashMap<String, String>();
+
+    /** Twitter streaming API endpoint example, '/statuses/filter.json' or '/statuses/firehose.json' */
+    private String endpointUrl;
+
+    /**
+     * Twitter streaming Host. This is paired with endpoint in @link{com.twitter.hbc.core.HttpHosts} constants.
+     * This is for configuring mock or enterprise endpoint.
+     */
+    private HttpHosts hosts;
+
+    /** OAuth params holder */
+    private OAuthSettings oAuthSettings;
+
+    private volatile boolean stopped;
+    private Integer bufferCapacity = 100000;
+    private Client client;
+
+    /** Process stream asynchronously */
+    private final ExecutorService tweetStreamProcessor = Executors.newSingleThreadExecutor();
+
+    private final String SITE_USER_ID_KEY = "follow";
+
+    public TwitterStreamer(OAuthSettings oAuthSettings){
+        this.oAuthSettings = oAuthSettings;
+    }
+
+    /**
+     * Starts streamer.
+     *
+     * @throws IgniteException If failed.
+     */
+    public void start() throws IgniteException {
+        if (stopped)
+            throw new IgniteException("Attempted to start an already started Twitter Streamer");
```

Comment:
Seems new client can be started instead of fail. "Already" started case should be checked instead.